src/content/en/updates/2018/06/audio-worklet-design-pattern.md

```diff
@@ -0,0 +1,414 @@
+thread. (See
+[this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Instance)
+for the more info.)
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-2.png">
+  <figcaption>
+    WASM module instantiation pattern B: Using AudioWorkletNode constructor's
+    cross-thread transfer
+  </figcaption>
+</figure>
+
+The pattern B can be useful if asynchronous heavy-lifting is required. It
+utilizes the main thread for fetching the glue code from the server and
+compiling the module. Then it will transfer the WASM module via the constructor
+of AudioWorkletNode. This pattern makes even more sense when you have to load
+the module dynamically after the AudioWorkletGlobalScope starts rendering the
+audio stream. Depending on the size of the module, compiling it in the middle of
+the rendering can cause glitches in the stream.
+
+Currently, the pattern B is only supported behind an experimental flag because
+it requires WebAssembly structured cloning.
+(`chrome://flags/#enable-webassembly`)
+If a WASM module should be a part of your AudioWorkletNode design, passing it
+through AudioWorkletNode constructor can definitely be useful.
+
+#### WASM Heap and Audio Data
+
+WebAssembly code only works on the memory allocated within a dedicated WASM
+heap. In order to take advantage of it, the audio data needs to be cloned back
+and forth between the WASM heap and the audio data arrays. The
+[HeapAudioBuffer](https://github.com/GoogleChromeLabs/web-audio-samples/blob/gh-pages/audio-worklet/design-pattern/lib/wa2.js#L34)
+class in the example code handles this operation nicely.
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-3.png">
+  <figcaption>
+    HeapAudioBuffer class for the easier usage of WASM heap
+  </figcaption>
+</figure>
+
+There is an [early proposal](https://github.com/WebAudio/web-audio-api/issues/1384)
+under discussion to integrate the WASM heap directly into the AudioWorklet
+system. Getting rid of this redundant data cloning between the JS memory and the
+WASM heap seems natural, but the specific details need to be worked out.
+
+### Handling Buffer Size Mismatch
+
+  - [Example code](https://googlechromelabs.github.io/web-audio-samples/audio-worklet/design-pattern/awn-ring-buffer/)
+
+An AudioWorkletNode and AudioWorkletProcessor pair is designed to work like a
+regular AudioNode; AudioWorkletNode handles the interaction with other codes
+while AudioWorkletProcessor takes care of internal audio processing. Because a
+regular AudioNode processes 128
+[frames](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API/Basic_concepts_behind_Web_Audio_API#Audio_buffers_frames_samples_and_channels)
+at a time, AudioWorkletProcessor must do the same to become a first-class
+citizen. This is one of the advantages of the AudioWorklet design that ensures
```

Comment:
"design that ensures no additional" -> "design: no additional"?

If you change this, then line 107 "but it can be" needs to be changed to be a complete sentence.  Maybe "However, it can be"?